KOGITO-277: Stunner - Enable support for Service Tasks (Work Item Handlers)

romartin opened the pull request:
Hey guys,

This PR provides support service tasks in the kogito space. See [KOGITO-277](https://issues.redhat.com/browse/KOGITO-277)

This PR is a continuation of @jesuino 's work ([see PR](https://github.com/kiegroup/kie-wb-common/pull/3034)), and it also includes the fixes for the [missing points](https://issues.redhat.com/browse/KOGITO-277?focusedCommentId=13936693&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13936693) previously identified on top. So it's also using the "new resouce api" from kogito to fetch the declarations and icons for service tasks, if present.

I'll keep working on adding tests and completing the TODO's tomorrow, but the `bpmn-kogito-runtime` webapp it's already running fine (or should be :)), so can you please to the review on the meanwhile??

@ederign @jesuino please can you build the `bpmn-kogito-runtime` and test the generated plugin work properly in VSCode? I'm just testing the default work items that are provided in the BC, also same images, put all those in the same path (.wid & images) and it'll properly import all those.

Thanks in advance!

romartin commented:
CC @porcelli @krisv @hasys  @inodeman

romartin commented:
Jenkins execute full downstream build

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -55,4 +61,67 @@ public WorkItemDefinitionStandaloneClientService(final Promises promises,
+    // TODO: Cache?
+    //  * Called once when loading process - (1) BPMNSessionInitializer
+    //  * Called twice when loading process - (1) BPMNClientDiagramService & (2) BPMNSessionInitializer
+    private Promise<Collection<WorkItemDefinition>> loadAllWorkItems() {
```
such TODO should be fixed or moved to Jira ticked

romartin replied:
done

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/test/java/org/kie/workbench/common/stunner/bpmn/client/workitem/WorkItemDefinitionClientParserTest.java:
```diff
@@ -0,0 +1,116 @@
+
+public class WorkItemDefinitionClientParserTest {
+
+    final static String WID = "  [\n" +
```
I guess it will be better to move it to separate resource file

romartin replied:
this class was just for testing, it has been removed

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/workitem/WorkItemDefinitionClientParser.java:
I guess it should be moved to Jira issues as well.

romartin replied:
yeah maybe a new JIRA...

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/KogitoStaticResourceContentService.java:
```diff
@@ -0,0 +1,61 @@
+    private static final Map<String, String> WID_ENTRIES =
+            Collections.singletonMap(BC_DEFAULT_URI, KogitoStaticWorkItemDefinitions.BC_DEFAULT);
+    private static final Map<String, String> ICON_ENTRIES =
+            new HashMap<String, String>() {{
```
It's better to not use this double curly braces initialization. From Sonar Cloud:

> Because Double Brace Initialization (DBI) creates an anonymous class with a reference to the instance of the owning object, its use can lead to memory leaks if the anonymous inner class is returned and held by other objects. Even when there's no leak, DBI is so obscure that it's bound to confuse most maintainers.

romartin replied:
Yeahh agree!!! But this class was just for testing, it has been removed :+1:

romartin commented:
Thanks @hasys @jesuino for your reviews!!

@jesuino just a Q about icons plz - the idea is that the icon shown in your screenshots above is the "default" one, in case the work item definition has no `icon` field specified. Otherwise, if you specify the icons and put them in the right path, they'll use the given ones. So I understand you didn't really added icons in the vscode path right? That's why all icons are the same one (the "default"). Or is it some bug? 

It's just calling the resource API also to fetch the icons, so it should work also on vscode, but I didn't really tested it, I just tested some custom envrionemnt....

Thanks in advance!

romartin commented:
@jesuino also - did u tested to save and reopen a process with some service tasks? all good too? Thx!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -55,4 +61,67 @@ public WorkItemDefinitionStandaloneClientService(final Promises promises,
+        });
+    }
+
+    // TODO: REMOVE
```
done

romartin commented:
Jenkins execute full downstream build

romartin commented:
@hasys @jesuino 

Added new commit with some new fixes and all TODOs implemented. It's now just missing to add a few tests, but you can feel free to build the kogito webapp and use it, if you need so, it should be working as expected.

Thanks!

jesuino commented:
hello @romartin 

I was able yes to create a process, add a custom WID, save and re-open it with the tasks and the mappings I made.

Regarding the icons: I didn't test with a custom icon, I just added the wid file and I could see the WIDs in editor.

Thanks!

domhanak commented:
@romartin @jesuino Could you please detail the build process(from PR to VSCode extension with these changes) for me or point me towards some doc? I would like to create some job that could automate it for us to speed up the test cycle in future.

domhanak commented:
jenkins retest this

hasys commented:
Hi @domhanak,

please, share your finding how to test this PR here or directly in Jira issue description. Thank you! :)

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/workitem/WorkItemDefinitionClientParser.java:
I think we should avoid a dependency to the given uberfire module due to the `Pair` class. I think we should use standard java map entry. For example on this line we can use 
```
Map.Entry<String, String> attrs = new AbstractMap.SimpleEntry<>("", "");
```

What do you think?

romartin replied:
good point - done

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/workitem/WorkItemDefinitionClientParser.java:
Shouldn't be `java.lang.String`? Asking because below we use `java.lang.Object`

romartin replied:
Good point @jomarko , it looks a bit confusing yeah. Since this point, in Stunner, the work item definitions were being parsed by the jbpm engine itself, so we're not worring about this data type conversions. 
But it's true that for client side we need this new parser implementation, so I'd been researching a bit and finally figured out, although doesn't seems much correct, that this code is correct. The `String` is the name given in the jbpm core for the `StringDataType`, also `java.lang.Object` is the name `ObjectDataType`, see [here](https://github.com/kiegroup/jbpm/blob/master/jbpm-flow/src/main/java/org/jbpm/process/core/datatype/impl/type/StringDataType.java#L57) and https://github.com/kiegroup/jbpm/blob/master/jbpm-flow/src/main/java/org/jbpm/process/core/datatype/impl/type/ObjectDataType.java#L121.
So thanks for the pointer, it was making lot of sense, but it seems that code is correct anyway :+1

domhanak commented:
@hasys I am preparing a job that should help with testing these kinds of PR.
Basically it does following:
1. Build stunner 
2. Compile `kie-wb-common-stunner-bpmn-kogito-runtime`
3. Copy the built client from `target/kie-wb-common-stunner-bpmn-kogito-runtime` to `kogito-tooling/packages/kie-bc-editors-unpacked/bpmn`
4. Build `kogito-tooling`
5. You should have a `*.vsix` file that you can use in VSCode, chrome-extension should be built for use aswell.

Now to test the WID integration, install the VSCode extension and check following:
* There are default WID's as in BC
* You can add a *.wid file to your project and it should be shown by the editor (if WID is valid)
* Create a simple diagram with WID, download it and import it to BC and vice-versa (it should be identical)

Hope it makes sense.

hasys commented:
Great, thank you @domhanak.

Also, please, be in touch with @ederign, if I correct such a job was already in progress. Am I correct @ederign?

domhanak commented:
@hasys Well my job works already but we do not have LTS node in our jenkins... :( So the build will fail.

Anyway, I tried manually, but I am unable to see any work items in palette. Could be that I built it wrong, but I tried couple of times with same results.

Scenario 1
* I added work item definitions to a project/folder in VSCode
* Created a completely new bpmn file and opened it
* I do not see any work items for choosing in palette

Scenario 2
* I created a simple process in business-central (start -> Email -> end)
* Downloaded it and added it to a project (default work item definitions are present in folder)
* Email workitem is displayed as normal task


Additional observations:
Every time I open a business-process (single click), VSCode opens it two times. One tab loads fast and shows diagram while the other keeps showing loading spinner @tiagobento Is that known issue?

ederign commented:
@domhanak yeah, @tiagobento is working right now in a more integrated process for this.  Please see KOGITO-759 and KOGITO-739.

@domhanak about the VSCode open twice this is the expected behaviour until we have a proper third part support on VSCode (the new API).

tiagobento commented:
@domhanak From what you described I think it's possible that you have two different VSCode extension versions installed.

romartin commented:
@domhanak yep both scenarios you mention are the basic ones, that should work. No worries, you did well, I'm just tyring to rebase and package again to test it, will let you know ASAP. Thanks for testing!

romartin commented:
Hey @domhanak 

I just tested the VSCode integration and it works properly for me - if no .wid file on the folder, no service tasks available, otherwise I can see and use them properly (forget the icons on the screenshot, it was just a test, but those are being displayed if you also deploy them on the folder):

![Screenshot from 2020-01-08 23-54-12](https://user-images.githubusercontent.com/4602417/72023652-ae974b00-3273-11ea-9c77-edea878a0d23.png)

What I did was mostly was you told...:

1. Build from root `kie-wb-common/kie-wb-common-stunner`as: `mvn clean install -DskipTests -Dgwt.compiler.skip=true`
2. Build the kogito process editor war, from root `kie-wb-common/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime` as: `mvn clean install -DskipTests`
3. Unzip the war generated at `target/kie-wb-common-stunner-bpmn-kogito-runtime.war` into `kogito-tooling/packages/kie-bc-editors-unpacked/bpmn`
4. Follow the intructions about building and running in VSCode from [here](https://github.com/kiegroup/kogito-tooling/blob/master/README.md)
5. In VSCode, once running the plugin and having a new window, I imported a folder where I've 2 files: the [process1.bpmn](https://gist.github.com/romartin/43ac8c70016cb32007b3a6387ba63c57) and the [WorkDefinitions.wid](https://gist.github.com/romartin/bd998c554a8efe372d475c1e3e21650d)

At this point, I can sucessfull open the process1 or create new processes, and the service tasks are working properly for me.

Can you please verify the steps and let me know some feedback plz? On the meanwhile I'll work on some missing test coverage... thanks!!

romartin commented:
Just rebased from master

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey @domhanak - I'd been adding more test cases in a new commit, and also tested again the editor in VSCode, by running the above steps, and it works for me, expect for one thing. So anyway please 
 would appreciate if you can move forward the testing for this, considering icons still not working for VSCode. This is starting to be urgent, thanks in advance!

@jesuino @ederign The only issue is about trying to get the icon data using the Resource API, in VSCode, as it does not allow me by just adding the images on the path, as expected. I see an error in the console as:

`[Extension Host] Error retrieving file defaultbusinessrulesicon.png: Error: cannot open file:///home/roger/development/temp/vscode-wid-tests/defaultbusinessrulesicon.png. Detail: File seems to be binary and cannot be opened as text (at /home/roger/development/romartin/kogito-tooling/packages/vscode-extension-pack-kogito-kie-editors/dist/extension/extension.js:5635:23)
`
Actualy the way it works on BC is by retrieveng the icon data from the server (as base64), because it comes from the VFS and it can be defined by the user. So for the Kogito version I'm expecting the same behavior - once having the icon path from the WID declaration, just retrieveng the icon data by using the Resource API... 

WDYT? Does it makes sense to improve the resource fetching mechanisms to achieve binary data reading? Otherwise, how/where should I obtain the icon data from?

Thanks!

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
also please @hasys @jomarko let us know some feedback on this PR, you're involved on the reviews.. thx in advance!

romartin commented:
Jenkins retest this

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/workitem/WorkItemDefinitionClientParser.java:
good point - done

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey @jomarko @domhanak - I've also included your code improvments suggested. Can you please give some feedback on the reviews? Thx!

Also @ederign @jesuino rather than the missing ability to load the icons data (see [above comment](https://github.com/kiegroup/kie-wb-common/pull/3087#issuecomment-572790668)), which maybe implies some change from fundation side too, some "default" icons appear anyway, so if you think it's fine we can go ahead merging this, once @jomarko @domhanak reviews are done, and then we can focus on the icons in a second iteration, it should be easy to fix and it allows us to support service tasks anyway, better than nothing. WDYT?

Thx!

jesuino commented:
Hello @romartin 

For now the API indeed does not support binary content, but we went ahead and created the following JIRA to address this. You can merge this PR and later I will send another one according to the changes introduced in the Resource Content API.

Thanks!

[1] https://issues.redhat.com/browse/KOGITO-862?page=com.atlassian.jira.plugin.system.issuetabpanels:all-tabpanel

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

domhanak commented:
Hi @romartin 

So I managed to get it working and now I see the service tasks in VSCode. 
The icons are not blocking, but I am unable to save the created process.
Is that known. I couldn't find JIRA for that?

romartin commented:
Hey @domhanak 

thanks for the feedback. About the multiple .wid files, yeah good point, lemme try to see if I can fix it.

About the issue that prevents you from saving, does it appears some error? any feedbacK I'll anyway try again now, but saving was initially working for me as well.

romartin commented:
Summarizing issues:
* Process is not saving - check it
* Support multiple `.wid` files on the project
* Icons not working

romartin commented:
Hey @domhanak @ederign @porcelli @jesuino 

Although this stuff probably works good enough for merging, there are 3 things remaining, I'd been working.. let's re-cap plz:
* Support multiple `.wid` files --> I've FIXED the issue found by @domhanak  and added a test case for it - now the editor will load properly all `.wid` files found in the project, not the just last one
* I've also tested the Save button  (also using Ctrl+S) and it works for me, although I'm using an old VSCode version I guess? (`1.36.1`), anyway it's properly saving for me. What's wrong @domhanak on your environment? if you open the browser console when trying to save, do you get some log? It works for me..
* Finally it's well known that icons are still not properly loaded - @jesuino is working on it ( see [KOGITO-862](https://issues.redhat.com/browse/KOGITO-862) )

So plz @domhanak @jomarko

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
@domhanak @jomarko @hasys Downstream build failed due some unrelated test, but artifacts are generated as well (except the kogito webapp you've to test here we're working on this, so you'll have to build it locally plz) - It's ready for review again

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/README.md:
```diff
@@ -11,6 +11,8 @@ the [Stunner directory README documentation](../../../).
 * Start GWT super dev mode by: `mvn gwt:run`
 * To create new diagram copy/paste this command into the browser console:
   * `gwtEditorBeans.get("BPMNDiagramEditor").get().setContent("")` 
+  * `window.frames.editorFrame.contentWindow.gwtEditorBeans.get("BPMNDiagramEditor").get().setContent("")`
```
We have a PR updating this method https://github.com/kiegroup/kie-wb-common/pull/3099

Where should I put note we do not forget update this README?

romartin replied:
what do you mean? sorry lost

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -55,4 +77,103 @@ public WorkItemDefinitionStandaloneClientService(final Promises promises,
+
+    private static boolean isNotEmpty(String s) {
+        return null != s && s.trim().length() != 0;
+    }
```
This PR defines same logic second time. At the same time we have the same in `org.kie.workbench.common.stunner.core.util.StringUtils` can we do something about it?

romartin replied:
will do it once possible, although this should not be blocking the PR plz, can you already approve?

romartin replied:
done

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
@ederign @tiagobento 
Hey guys, me and @domhanak are not able to build and run the VSCode plugin locally, as we're doing before. 
I noticed that the build now depends on the WAR artifacts for the editors, but although bulding those locally we're not able to run the plugin properly anymore.
Please that's really important for moving this stuff forward, please can you let us know how to proceed??
Thx

tiagobento commented:
@romartin Hi Roger. We modified the build to fetch the editor WARs using Maven, so take a look at `packages/kie-bc-editors-unpacked/pom.xml` and see if the dependency that's there is the same dependency that you're using. Pere warned me yesterday that the names changed from version `7.32.0-SNAPSHOT` to `7.33.0-SNAPSHOT`, so I think you will have to change something there to test it locally. Right now, to build the VSCode you don't have to do any manual proccess, just building it with yarn run build:prod should give you a .vsix file.

ederign commented:
@tiagobento just make sure that we have already published 7.33 artifacts on maven central. (probably our ci will identify this)

romartin commented:
@ederign @tiagobento I'm building the artifacts locally anyway for testing

romartin commented:
Hey @domhanak @jomarko 
Please rememer to provide some feedback here, are missing your final reviews. 
Thanks!

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

manstis commented:
Jenkins please execute full downstream build.

romartin commented:
Jenkins please execute full downstream build

romartin commented:
@mbiarnes @manstis no way to make the FDB work.... :( any ideas?

mbiarnes commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

mbiarnes commented:
Jenkins execute full downstream build

mbiarnes commented:
Jenkins execute full downstream build

romartin commented:
@domhanak please builds are FINALLY green!! can we provide some final review status?? Thanks!

romartin commented:
@domhanak ping

romartin commented:
Hey @LuboTerifaj - just noticed @domhanak is on PTO, but his review is still missing. Can you handle that? Thx in advance

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
rebased again - please @domhanak only missing your final review, please lemme know something. Thanx!

romartin commented:
@domhanak all green again! thx!

romartin commented:
Jenkins retest this

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey @domhanak 

Sorry, now when testing it again, I noticed I didn't told you before, but that's kinda "known issue". I suspect that the process is not being saved, once having some service task, not because some error , just because you should be missing to fill some field (so validation is not success). Please check the console and also try, for example, saving a process with a `Log` task, both without the `message` field set and empty. You'll see the results... 

The "problem" described is something we should fix urgently too, but it's a different topic - how editors expose the messages for each channel, otherwise we can end up in this kinda situation. (it's not your fault @domhanak , I also faced this some time before =) )

Soo IMO this is more than ready to merged guys, it's starting to be really old and urgent. WDYT @domhanak , just missing your final review!! 

CC @porcelli @ederign 

Thanks!

romartin commented:
rebased and squahsed

domhanak commented:
@romartin Ok, issue reported here - https://issues.redhat.com/browse/KOGITO-1076

domhanak commented:
jenkins retest this
